### PR TITLE
Localize setup screen

### LINKED ISSUE #3117: Setup: Take culture into account
I create an OC instance (I test with the source code) with the Saas recipe, enable Localization, add fr-FR culture and select it in the settings, create src\OrchardCore.Cms.Web\App_Data\Localization and add OrchardCore.Admin-fr.po and OrchardCore.Setup-fr.po.

When I go to /admin, the strings are localized.

When I create a tenant and go to the setup, I expect that the strings are localized but they are not, even if I add in the querystring `&culture=fr-fr`

### PR DESCRIPTION
**Pull request by hishamco**
Fixes #3117

### CODE REVIEW

**Comment by hishamco** [comment]
@agriffard could you please retry your earlier scenario with my changes and let me know your feedback

**Comment by agriffard** [comment]
It works but not for the first setup.

When I copy .po files in App_Data\Localization and add explicitly the `fr` culture and set it as the default one, then I can create a tenat whose setup screen is localized:
image

Important note: It does not mean that the created tenant will be created with this culture though.
We will need some kind of 'Add and Set culture' recipe step that takes the current culture of the setup screen into account.

**Comment by hishamco** [comment]
Thanks for your feedback, that's why I mentioned earlier in the original issue, the first installation will not be localized unless we define a preset of supported cultures (may be the requested once)

Again my opinion to have a drop down list that shows the cultures in setup instead of injecting the culture in querystring

**Comment by hishamco** [comment]
> We will need some kind of 'Add and Set culture' recipe step that takes the current culture of the setup screen into account.

May be but again from user experience perspective we need to list the setup cultures if there are many

**Comment by hishamco** [comment]
Again if each recipe has setup step that defines a set of supported cultures, this will be invalid because we don't know which the user pick. Setup language is somethings related to the application regardless of what recipes there.

IMHO `OrchardCore.Cms.Web\appsettings.json` is a suitable place for define such thing

**Comment by hishamco** [comment]
@agriffard I added the localization support for the default tenant using `appsettings.json`, @sebastienros the setup screen is now localizable for the default installation or tenant creations.

In case you dislike the `appsettings.json` - which is I think the better choice - please let me hear from you to provide supported cultures for the first installation

**Comment by agriffard** [comment]
Ok, I added the fr culture in appsettings.json: 
`"SupportedCultures":  [ "en", "fr" ]`
Then, I can translate the setup screen on first install by adding `?culture=fr` to the querystring

**Comment by hishamco** [comment]
Backing to my earlier suggestion why we didn't provide a drop down list to the user to choose the setup language instead of playing with the query string, coz I didn't like the current behavior from the user experience perspective

**Comment by hishamco** [comment]
@sebastienros your thoughts

**Comment by Skrypt** [comment]
It could easily be done with CSS + javascript that would display a culture selector on first setup page then post the result to the same page + url param for culture.

**Comment by hishamco** [comment]
> It could easily be done with CSS + javascript

Yep we can show up a modal before the setup start or adding a simple drop down list in the setup page that inject the proper culture querystring if the language selector has changed

**Comment by Skrypt** [comment]
I like the modal idea

**Comment by hishamco** [comment]
Waiting for Seb thought, meanwhile I can prototype the language selector modal ..

**Comment by hishamco** [comment]
The language selector looks like this

! [Setup1](httos://user-images.githubusercontent.com/3237266/56851356-b944c680-6916-11e9-878b-94444b67198c.png)

! [Setup2](https://user-images.githubusercontent.com/3237266/56871132-33aed00-6a22-11e9-8e4b-ec577a10d623.png)

or even better with flags

[Setup3](https://user-images.githubusercontent.com/3237266/56871142-50e30b80-6a22-11e9-8f30-3c7333cf0111.png)

**Comment by sebastienros** [comment]
Thanks Hisham for this. After looking at your changes I would suggest this:

Don't require Localization (and Settings) modules from Setup.
Instead, copy from the Localization module what is necessary, which are:
- copy the code from ConfigureServices from the Localization module
- copy the code from Configure from the Localization module, without the Settings module part

Then like you did we can configure the default and supported cultures directly in the appsettings.

Because it's a Setup module it's totally ok to just rely on the app settings for this. Then it will use the standard aspnet logic to resolve which culture to use (querystring, request, default). And in most case it will come from the client's browser.

Later we think about a front-end culture picker like you suggested. And we'll definitely have one with the Localization module.

**Comment by hishamco** [comment]
I like your suggestions @sebastienros but what's the wrong with current implementation? Because seems We are cheating here 😊 .. let me think and let me knew if you have a better idea

> Later we think about a front-end culture picker like you suggested. And we'll definitely have one with the Localization module.

You read my mind 😊 yesterday I'm think to create a culture picker in the localisation module, this will be useful for both front and back-end

**Comment by sebastienros** [comment]
My concern is that Localization requires Settings which should be initialized by the Setup. The Setup should not require Settings. It should take its settings from the configuration only.

This means there will be some code duplication, but it won't hurt. Setup is a special module.

**Comment by hishamco** [comment]
> This means there will be some code duplication, but it won't hurt

But I think this will make localization middleware works all the time. Am I right?

**Comment by sebastienros** [comment]
No. it's just for the Setup module. Once the tenant is created the Setup module is not used anymore. It's only used by a temporary fake tenant (including services and middleware pipeline)

**Comment by hishamco** [comment]
Ok .. I will do the needed changes now

**Comment by sebastienros** [comment]
And please fix the build that I have broken :/

**Comment by hishamco** [comment]
Sure

File: src/OrchardCore.Modules/OrchardCore.Localization/Startup.cs [hunk 1]
```diff
@@ -15,6 +17,17 @@ namespace OrchardCore.Localization
     /// </summary>
     public class Startup : StartupBase
     {
+        public Startup(IHostingEnvironment env)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Localization/Startup.cs]: Unused, and not necessary

File: src/OrchardCore.Modules/OrchardCore.Localization/Startup.cs [hunk 2]
```diff
@@ -43,11 +55,12 @@ public override void Configure(IApplicationBuilder app, IRouteBuilder routes, IS
                     .ToArray();
 
                 options
-                    .AddSupportedCultures(supportedCulture)
-                    .AddSupportedUICultures(supportedCulture);
+                    .AddSupportedCultures(siteSettings.SupportedCultures)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Localization/Startup.cs]: wrong change?

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Localization/Startup.cs]: Oops .. it's my bad :(

**Comment by hishamco** [comment]
I fixed the old build, but i'm surprising now the tests passed on my end!!

File: src/OrchardCore.Modules/OrchardCore.Setup/Startup.cs [hunk 3]
```diff
@@ -1,20 +1,59 @@
...
+using Microsoft.Extensions.Options;
 using OrchardCore.Modules;
 
 namespace OrchardCore.Setup
 {
     public class Startup : StartupBase
     {
+        public Startup(IHostingEnvironment env)
+        {
+            var builder = new ConfigurationBuilder()
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Setup/Startup.cs]: @jtkech do we need this code to be able to read the configuration from the main appsettings.json? I don't think so.

@hishamco it should work by resolving `IShellConfiguration` directly. Look for an example in the Media module.

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.Setup/Startup.cs]: Normally not when resolving `IShellConfiguration` but better to try it in the context of the setup module.

Then, you don't need to resolve the `OrchardCore` section.

File: src/OrchardCore.Cms.Web/appsettings.json [hunk 4]
```diff
@@ -8,7 +8,8 @@
     }
   },
   "OrchardCore": {
-
+    "DefaultCulture": "",
+    "SupportedCultures":  [ "en" ]
```
Comment (jtkech) [inline src/OrchardCore.Cms.Web/appsettings.json]: "OrchardCore": {
      "OrchardCore.Setup": {
        "DefaultCulture": "",
        "SupportedCultures": [ "en" ]
      }

Note: If `DefaultCulture` is an empty string, at runtime you will get a null value.

File: src/OrchardCore.Modules/OrchardCore.Setup/Startup.cs [hunk 5]
```diff
@@ -1,20 +1,60 @@
...
+using Microsoft.Extensions.Configuration;
 using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
 using OrchardCore.Modules;
 
 namespace OrchardCore.Setup
 {
     public class Startup : StartupBase
     {
+        public Startup(IHostingEnvironment env)
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Setup/Startup.cs]: Didn't try the whole branch but just tried it by only copy paste the startup.

It works by just injecting an `IShellConfiguration configuration`.

File: src/OrchardCore.Modules/OrchardCore.Setup/Startup.cs [hunk 6]
```diff
@@ -1,20 +1,60 @@
...
+        {
+            var builder = new ConfigurationBuilder()
+                .SetBasePath(env.ContentRootPath)
+                .AddJsonFile("appsettings.json");
+
+            Configuration = builder.Build();
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Setup/Startup.cs]: So, just `Configuration = configuration;`

File: src/OrchardCore.Modules/OrchardCore.Setup/Startup.cs [hunk 7]
```diff
@@ -1,20 +1,60 @@
...
+            var options = serviceProvider.GetService<IOptions<RequestLocalizationOptions>>().Value;
+            var defaultCulture = Configuration["OrchardCore:OrchardCore.Setup:DefaultCulture"];
+            var supportedCultures = Configuration.GetSection("OrchardCore:OrchardCore.Setup:SupportedCultures").Get<string[]>();
+
+            if (!string.IsNullOrEmpty(defaultCulture))
+            {
+                options.SetDefaultCulture(defaultCulture);
+            }
+
+            if (supportedCultures.Length > 0)
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Setup/Startup.cs]: Maybe a null check if nothing is provided in the config.

    if (supportedCultures?.Length > 0)

File: src/OrchardCore.Modules/OrchardCore.Setup/Startup.cs [hunk 8]
```diff
@@ -1,20 +1,60 @@
...
+            services.AddPortableObjectLocalization(options => options.ResourcesPath = "Localization");
             services.AddSetup();
-		}
+        }
 
-		public override void Configure(IApplicationBuilder builder, IRouteBuilder routes, IServiceProvider serviceProvider)
+		public override void Configure(IApplicationBuilder app, IRouteBuilder routes, IServiceProvider serviceProvider)
         {
+            var options = serviceProvider.GetService<IOptions<RequestLocalizationOptions>>().Value;
+            var defaultCulture = Configuration["OrchardCore:OrchardCore.Setup:DefaultCulture"];
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Setup/Startup.cs]: So, without the `OrchardCore` section and maybe a null check if it is intended to use the invariant culture if the default culture is an empty string in the `appsettings.json`. Anyway, below we check `IsNullOrEmpty`, so as you want.

    var defaultCulture = Configuration["OrchardCore.Setup:DefaultCulture"] ?? String.Empty;

and then

    var supportedCultures = Configuration.GetSection("OrchardCore.Setup:SupportedCultures").Get<string[]>();

**Comment by jtkech** [comment]
You need an `OrchardCore.Setup` section in your `appsettings.json`.
Then maybe a null check ` if (supportedCultures?.Length > 0)`.
See my review.

File: src/OrchardCore.Cms.Web/appsettings.json (hunk 4 above)
Comment (hishamco) [reply src/OrchardCore.Cms.Web/appsettings.json]: No problem because we check for null or empty in the localization middleware

Comment (jtkech) [reply src/OrchardCore.Cms.Web/appsettings.json]: Yes for `DefaultCulture` but not done for `SupportedCultures`. I saw it failing when running unit tests and it was working with `if (supportedCultures?.Length > 0)`.

Then, to resolve your config values you need an `OrcharCore` section but also an `OrchardCore.Setup` sub section in your `appsettings.json`.

Comment (hishamco) [reply src/OrchardCore.Cms.Web/appsettings.json]: > Then, to resolve your config values you need an OrcharCore section but also an OrchardCore.Setup sub section in your appsettings.json.

It's already there in the Web project

Comment (jtkech) [reply src/OrchardCore.Cms.Web/appsettings.json]: As i can see in appsettings.json there is only the `OrchardCore` section.

Yes, in any case we need the OC section, then we use a module sub section if we want config values that are specific to a module.

Then, by using `IShellConfiguration` at runtime you don't need to specify the OC section but still the module sub section if there is one.

**Comment by hishamco** [comment]
> You need an OrchardCore.Setup section in your appsettings.json.

You made me crazy 😂 coz what's you asking for is trivial .. seems it's my fault that I didn't push the changes of this file while I demoing the culture selector

thanks

**Comment by jtkech** [comment]
no problem ;)

File: src/OrchardCore.Modules/OrchardCore.Setup/OrchardCore.Setup.csproj [hunk 9]
```diff
@@ -6,10 +6,12 @@
 
   <ItemGroup>
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.DisplayManagement\OrchardCore.DisplayManagement.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Localization.Core\OrchardCore.Localization.Core.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Module.Targets\OrchardCore.Module.Targets.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Data\OrchardCore.Data.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Recipes.Abstractions\OrchardCore.Recipes.Abstractions.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Setup.Core\OrchardCore.Setup.Core.csproj" />
+    <ProjectReference Include="..\..\OrchardCore.Modules\OrchardCore.Localization\OrchardCore.Localization.csproj" />
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Setup/OrchardCore.Setup.csproj]: No reference to other modules allowed.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Setup/OrchardCore.Setup.csproj]: If is that .. that mean we have a full clone not a small duplicate code as you mentioned earlier 😂if it's fine I need to copy everything to make things work as it's now

**Comment by hishamco** [comment]
Any updates on this one 🕐📆

**Comment by hishamco** [comment]
Just reminder this should merged after merging the PR #3732

**Comment by hishamco** [comment]
⌚️ 📆

**Comment by hishamco** [comment]
@sebastienros I don't think there's any extra things we need to add here

**Pull request merged**